plugins/yara/src/plugin.cpp

```diff
@@ -0,0 +1,423 @@
+    switch (status) {
+      default:
+        die("unhandled result value for yr_rules_scan_mem");
+      case ERROR_SUCCESS:
+        break;
+      case ERROR_INSUFFICIENT_MEMORY:
+        return caf::make_error(ec::unspecified,
+                               "insufficient memory to scan bytes");
+      case ERROR_TOO_MANY_SCAN_THREADS:
+        return caf::make_error(ec::unspecified, "too many scan threads");
+      case ERROR_SCAN_TIMEOUT:
+        return caf::make_error(ec::unspecified, "scan timeout");
+      case ERROR_CALLBACK_ERROR:
+        return caf::make_error(ec::unspecified, "callback error");
+      case ERROR_TOO_MANY_MATCHES:
+        return caf::make_error(ec::unspecified, "too many matches");
+    }
+    return builder.finish_as_table_slice("yara.match");
+  }
+
+private:
+  static auto callback(YR_SCAN_CONTEXT* context, int message,
+                       void* message_data, void* user_data) -> int {
+    TENZIR_ASSERT(user_data != nullptr);
+    auto* builder = reinterpret_cast<series_builder*>(user_data);
+    if (message == CALLBACK_MSG_RULE_MATCHING) {
+      auto* rule = reinterpret_cast<YR_RULE*>(message_data);
+      TENZIR_DEBUG("got a match for rule {}", rule->identifier);
+      YR_STRING* string = nullptr;
+      yr_rule_strings_foreach(rule, string) {
+        auto row = builder->record();
+        auto rec = row.field("rule").record();
+        rec.field("identifier").data(rule->identifier);
+        rec.field("namespace").data(std::string_view{rule->ns->name});
+        auto rule_string
+          = std::string_view{reinterpret_cast<const char*>(string->string),
+                             detail::narrow_cast<size_t>(string->length)};
+        rec.field("string").data(rule_string);
+        const char* tag = nullptr;
+        auto tags = rec.field("tags").list();
+        yr_rule_tags_foreach(rule, tag) {
+          tags.data(std::string_view{tag});
+        }
+        auto meta_rec = rec.field("meta").record();
+        YR_META* meta = nullptr;
+        yr_rule_metas_foreach(rule, meta) {
+          auto identifier = std::string_view{meta->identifier};
+          if (meta->type == META_TYPE_INTEGER)
+            meta_rec.field(identifier).data(int64_t{meta->integer});
+          else if (meta->type == META_TYPE_BOOLEAN)
+            meta_rec.field(identifier).data(meta->integer != 0);
+          else
+            meta_rec.field(identifier).data(std::string_view{meta->string});
+        }
+        YR_MATCH* match = nullptr;
+        auto list = rec.field("matches").list();
+        yr_string_matches_foreach(context, string, match) {
+          auto match_rec = list.record();
+          match_rec.field("identifier").data(string->identifier);
+          auto bytes = std::span<const std::byte>{
+            reinterpret_cast<const std::byte*>(match->data),
+            detail::narrow_cast<size_t>(match->data_length)};
+          // TODO: switch to blob type once available.
```

Comment:
That now exists.

Reply:
Since YARA rules are also frequently used for plain text, I think we should add an option to display this either as string or as bytes. Default to bytes, but opt-in via `--string-data` to use a string. Otherwise `"bar"` gets represented as `YmFy` in Base64, which may not be helpful at all.

Reply:
You could also do UTF-8 validation for it (simdjson should have something for this) to decide this automatically.

Reply:
Uh, that feels like too much magic for me. A user would want to process *either* downstream.

Reply:
> Or perhaps better: make that an option in the respective printer?

I'd like to hear @jachris' thoughts on this. I'm not entirely opposed to it, but we would need to maintain that for every individual printer.

Reply:
Another option: just make this a function so that users can do `yara <args> | replace data=f(data)` for an appropriately named function `f`.

For now, I'll just use the `blob` type as it's the most correct one.

Reply:
Makes sense to me. We discussed something similar a while ago for `--numeric-durations`, which is an option for the JSON printer currently. That, too, should be a function that converts from duration to double.

Reply:
43f5ce853f has the delta. For our text example, it's no longer optimal, but it should work.

Reply:
I tend to value consistency a lot, but I guess it depends on what people usually do afterwards with the data. It would be quite surprising to suddenly change types when you don't expect it, so this should be opt-in if anything. I'm not against providing additional encoding options for JSON. We could also encode something like `b"hello\xFF"` as `"hello\\xFF"` in that case (double backslash is required), but base64 still seems to be the most common solution.

> replace data=f(data)

Agreed, this would be similar to `data = data.decode()` in Python (which can fail).

Reply:
> We could also encode something like `b"hello\xFF"` as `"hello\\xFF"` in that case (double backslash is required), but base64 still seems to be the most common solution.

Indeed, Base64 is the fallback. But I've seen "hex-encoding" pretty frequently. Zeek logs, for example, always use that instead. As a user of those, I find it highly valuable to see the ASCII characters and only "get blinded" for non-printable bytes.

@jachris what do you think of defaulting to Base64 and provide hex-encoding as an opt-in?

Reply:
For Zeek it seems to be hex-encoding, but that actually leads to a lot of trouble since we would need to validate whether it's UTF-8 when parsing. Suricata by default encodes binary payloads as base64.

Ultimately, I see this as yet another use-case for functions. We should still agree on a default, and ideally before the next release when we include the blob type.

NB: I unresolved this since we're actively discussing still, the thread is hard to find otherwise.

Reply:
We always use hex enconding that for Zeek logs, as you can see [here](https://github.com/tenzir/tenzir/blob/bca0cd07f8b6c619a8e01245ca2f7180eefc2150/libtenzir/builtins/formats/zeek_tsv.cpp#L404-L421). 

But for JSON, I don't think there is a common way to do this besides base64. We can of course add it anyways, but this is what makes me hesitate. Or have you seen JSON like this in practice?

For the explorer, this is also not a problem. We can of course show `b"hello\xFF"` there (see https://github.com/tenzir/app/issues/333).

Reply:
>But for JSON, I don't think there is a common way to do this besides base64. We can of course add it anyways, but this is what makes me hesitate. Or have you seen JSON like this in practice?

No, which is why I think the current default is good.

I am resolving this thread, as the discussion should probably take place in a different issue that's more discoverable.